## **Комментарий ревьюера**

Привет 👋

Меня зовут Валентин, и я буду проверять твой проект. Предлагаю сразу договориться общаться на «ты», если для тебя это не уместно, то дай, пожалуйста, знать, будем общаться на «вы».

### **Краткое резюме по ревью**

<div style="border-radius: 15px; box-shadow: 2px 2px 4px; border: 1px solid; background:Gainsboro; padding: 20px">
        
Видны твои старания при работе над данным проектом. Большинство основных этапов в работе сделано хорошо:
- есть много сильных и интересных алгоритмов 🤩
- детально изучены многие тонкости данных
- здорово сделана предобработка данных
- качественно исследованы данные
- сделано много верных промежуточных выводов, основанных на полученных данных

Твоего внимания требует:
- пересмотр вывода

Я описал замечания в комментариях, которые помечены цветом(красным, желтым), после их доработки я смогу принять работу.
    
Спасибо тебе за твой труд! Я на все 100% уверен, что у тебя все получится, и с нетерпением буду ждать твою работу на повторное ревью 😉

</div>

### **PS:**

Хочу поделиться дополнительным материалом, он бесплатный, но объемный(может уйти несколько месяцев):
- У Тимофея Хирьянова есть [Плей-лист по алгоритмам на python](https://www.youtube.com/playlist?list=PLRDzFCPr95fK7tr47883DFUbm4GeOjjc0), это вроде просто видео-лекции, но в описании есть ссылка на лекции и практику, не стесняйся смотреть, преподаватель очень наглядно объясняет материал. На первых уроках задания могут показаться сложными, так как требует знания геометрии, не стесняйся проходить далее, будет проще)

### **Правила оформления:**

По ходу тетрадки ты найдешь комментарии к проекту, они будут в отдельных ячейках Markdown с заголовком «Комментарий ревьюера». Постарайся учесть эти комментарии для дальнейших проектов. Я использую заголовки в комментариях, они изменят твое оглавление. Но тебе будет удобно перемещаться от комментария к комментарию. А перед перемещением проекта в портфолио просто уберешь данные комментарии.

Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта быстрее.

Я буду красить комментарии цветом:

<div class="alert alert-danger"; style="border-left: 7px solid FireBrick">
<h4> ❌ <font color="FireBrick">Комментарий ревьюера (красный)</font></h4>

Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.

Также жду от тебя комментарии, где вкратце опишешь изменения, которые были внесены.
</div>

<div class="alert alert-warning"; style="border-left: 7px solid Gold">
<h4> ⚠️ <font color="DarkGoldenRod">Комментарий ревьюера (желтый)</font></h4>

Такой комментарий является рекомендацией или советом. Можешь использовать их на своё усмотрение.
    
В случае изменений кода, опиши, пожалуйста, вкратце то, что изменилось. Если изменений не было, то просто опиши свои мысли на эту тему.
</div>

<div class="alert alert-success"; style="border-left: 7px solid green">
<h4> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h4>

Такой комментарий говорит о том, что было сделано что-то качественное и правильное 😃
    
Я не буду ждать твоего комментария, но буду ему очень рад 🤗
</div>

Чтобы мы были на одной волне, рекомендую общаться в том же стиле 😊 Твои ответы на мои комментарии будут подкрашиваться в синий цвет.

Ниже я оформил образец для твоего удобства, тебе остается только добавить после моего комментария ячейку с типом **Markdown**, скопировать туда код(чтобы увидеть код дважды, кликни на ячейку) из образца и написать свои мысли по поводу комментария.

<div class="alert alert-info"; style="border-left: 7px solid SteelBlue">
<h5>🎓 <font color="SteelBlue">Комментарий от студента (синий)</font></h5>    

Привет Валентин! 
- В первую очередь благодарю за оценку моего проекта, и коментариям к нему!  
- Отдельное спасибо за предоставленный доп-материал, который непременно возьму на изучение!

</div>

## **Комментарий ревьюера v2**

### **Краткое резюме по ревью**
<div style="border-radius: 15px; box-shadow: 2px 2px 4px; border: 1px solid; background:Gainsboro; padding: 20px">

Молодец, исправлено:
- пересмотр вывода

Спасибо тебе за твой труд! Я готов принять работу. Успехов в дальнейшем обучении!

</div>

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Перед исследованием зависимости между количеством детей и способностью кредитора выполнять свои обязательства по кредиту,
сформулирую некоторые тезисы.
1) Люди без детей, менее ответственны к обязательствам по кредиту. Как правило, это категория одиноких граждан, либо молодых!
   Данная категория, скорее более легкомысленная, безответственная, в целом импульсивная и менее склонна к планированию.
   Люди которые скорее хотят всего и сразу, без дальнейших последствий.

2) Люди с детьми (1 или 2 ребенка), более ответственно относятся к своим обязательствам по кредиту.
   Данная категория здравомыслящая, ответственная, планирующая будущее для своих детей.

3) Семьи с 3 и более детьми, самые безответственные по кредитным обязательствам.
   Многодетные семьи имеют минимальный доход на одного члена в семье. Категория имеет больший риск непредвиденных расходов.
   
 Проверка тезисов, даст нам необходимый вектор для дальнейших исследований.



In [22]:
# Определяю долю просрочки по кредиту среди людей имеющих детей и бездетных.

debt_no_children = round(100*data['debt'].loc[data['children'] == 0].mean(),2)
debt_with_children = round(100*data['debt'].loc[data['children'] > 0].mean(),2)
print('Доля просрочки по кредиту категории граждан без детей:', debt_no_children, '%' )
print('Доля просрочки по кредиту категории граждан с детьми:', debt_with_children, '%' )

Доля просрочки по кредиту категории граждан без детей: 7.54 %
Доля просрочки по кредиту категории граждан с детьми: 9.24 %


Вывод: Мои утверждения оказались не верными ;(
       Категория граждан без детей, оказалась более ответственна к обязательствам по кредиту.

In [23]:
# Полная таблица зависимости количества детей к просрочке по кредиту, выглядит следующим образом:

for i in range(6):
    print('Доля просрочки граждан у которых',i,'детей, составляет:',
    round(100*data['debt'].loc[data['children'] == i].mean(),1),'%')

Доля просрочки граждан у которых 0 детей, составляет: 7.5 %
Доля просрочки граждан у которых 1 детей, составляет: 9.2 %
Доля просрочки граждан у которых 2 детей, составляет: 9.5 %
Доля просрочки граждан у которых 3 детей, составляет: 8.2 %
Доля просрочки граждан у которых 4 детей, составляет: 9.8 %
Доля просрочки граждан у которых 5 детей, составляет: 0.0 %


Последняя строка таблицы с долей в 0%, скорее всего говорит об отсутствии статистики для граждан с 5ю детьми.

Создам в датафрейме дополнительный столбец для категоризации по количеству детей.

Создам функцию категоризации по детям.

In [24]:
# Создаю функцию категоризазии по детям 
'''есть дети - 1 или 2 ребенка, 
   многодетный - более 3 детей'''
 
def children_category(children):
    if 1 <= children <= 2:
        return 'есть дети'
    if children >= 3:
        return 'многодетный'
    return 'нет детей'

In [25]:
# Создаю столбец children_state

data['children_state'] = data['children'].apply(children_category)

In [26]:
# Проверяю результат.

data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose total_income_category          purpose_category  \
0               покупка жилья                     B  операции с недвижимостью   
1     приобретение автомобиля                     C    операции с автомобилем   
2               покупка жилья                     C  операции с недвижимостью   
3  дополнительное образование                     B     получение образования   
4             сыграть свадьбу                     C        проведение свадьбы   

  children_state  
0      есть дети  
1      есть дети  
2      нет детей  
3    многодетный  
4      нет детей

In [27]:
# Для упрощения исследования, создаю таблицу с нужными мне столбцами.

debt_children = data[['children', 'debt', 'children_state']]

In [28]:
# Проверяю результат.

debt_children.tail()

children  debt children_state
21520         1     0      есть дети
21521         0     0      нет детей
21522         1     1      есть дети
21523         3     1    многодетный
21524         2     0      есть дети

Создаю сводную таблицу. 

In [29]:
# Создаю таблицу методом pivot_table().

debt_children = data.pivot_table(index='children_state', columns='debt', values='gender', aggfunc='count')

In [30]:
# Создаю столбцы таблицы.

debt_children.columns = ['no_debt', 'debt']

In [31]:
# Считаю [долю должников].

debt_children['share_of_debt'] = debt_children['debt'] / (debt_children['debt'] + debt_children['no_debt'])

In [32]:
# Перевожу значения в столбце [доля дожников] в проценты.

debt_children['share_of_debt'] = debt_children['share_of_debt'].map('{:.1%}'.format)

In [33]:
# Сортирую по столбцу [доля должников], по убыванию.

debt_children.sort_values(by='share_of_debt', ascending=False)

no_debt  debt share_of_debt
children_state                             
есть дети          6222   638          9.3%
многодетный         349    31          8.2%
нет детей         13028  1063          7.5%

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h5>

Да, у заемщиков с детьми больше вероятность просрочки

</div>

Вывод: Из таблиц видим следующий результат: категория граждан у которых есть дети, наиболее подвержена риску по кредитным обязательствам. С увеличение количества детей, увеличивается количество просроченных задолженностей по кредиту.
Причем семьи с 3 детьми, более ответственно относятся к кредитным обязательствам, чем семьи с 1 или 2 детьми.


#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [34]:
# Для упрощения исследования, создаю таблицу с нужными мне столбцами.

debt_family_status = data[['children', 'debt', 'children_state', 'family_status']]

In [35]:
# Проверяю результат.

debt_family_status.tail(10)

children  debt children_state     family_status
21515         1     1      есть дети   женат / замужем
21516         0     0      нет детей   женат / замужем
21517         0     0      нет детей  гражданский брак
21518         0     0      нет детей   женат / замужем
21519         1     0      есть дети         в разводе
21520         1     0      есть дети  гражданский брак
21521         0     0      нет детей   женат / замужем
21522         1     1      есть дети  гражданский брак
21523         3     1    многодетный   женат / замужем
21524         2     0      есть дети   женат / замужем

In [36]:
# Определяю долю просрочки по кредиту в категориях [family_status].

for i in data['family_status'].unique():
    print('Доля просрочки по кредиту в категории "',i,'"   составляет:',
    round(100*data['debt'].loc[data['family_status'] == i].mean(),1),'%')

Доля просрочки по кредиту в категории " женат / замужем "   составляет: 7.6 %
Доля просрочки по кредиту в категории " гражданский брак "   составляет: 9.3 %
Доля просрочки по кредиту в категории " вдовец / вдова "   составляет: 6.6 %
Доля просрочки по кредиту в категории " в разводе "   составляет: 7.1 %
Доля просрочки по кредиту в категории " Не женат / не замужем "   составляет: 9.8 %


Самыми надежными категориями по кредитным обязательствам, являются граждане " в разводе " и " вдовец / вдова ", 
наименее надежными кредиторами являются граждане состоящие в категории " гражданский брак " и " не женат / не замужем ", 
медианное значение за категорией " женат / замужем ".

Создаю сводную таблицу.

In [37]:
# Создаю таблицу методом pivot_table()

debt_family_status = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count')

In [38]:
# Создаю столбцы таблицы.

debt_family_status.columns = ['no_debt', 'debt']

In [39]:
# Считаю [долю должников].

debt_family_status['share_of_debt'] = debt_family_status['debt'] / (debt_family_status['debt'] + 
                                                                    debt_family_status['no_debt'])

In [40]:
# Перевожу значения в столбце [доля дожников] в проценты.

debt_family_status['share_of_debt'] = debt_family_status['share_of_debt'].map('{:.1%}'.format)

In [41]:
# Сортирую по столбцу [доля лолжников] по убыванию.

debt_family_status.sort_values(by='share_of_debt', ascending=False)

no_debt  debt share_of_debt
family_status                                     
Не женат / не замужем     2523   273          9.8%
гражданский брак          3749   385          9.3%
женат / замужем          11334   927          7.6%
в разводе                 1105    84          7.1%
вдовец / вдова             888    63          6.6%

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h5>

Правильно, заемщики состоящие или состоявшие в браке реже допускают просрочку

</div>

**Вывод:** Не женатые и не замужние, а так же граждане состоящие в гражданском браке, более всего подвержены риску долга по кредитным обязательствам. Могу предположить, что это связано с меньшей ответственностью за семейные ценности. Граждане данных категорий не имеют совместных семейных накоплений и более развязаны по финансам, что в итоге может привести к неплатежам по кредиту.


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Категоризация по уровню дохода:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [42]:
# Создаю таблицу с нужными мне столбцами.

debt_total_income_category = data[['debt', 'total_income', 'total_income_category']]

In [43]:
# Проверяю результат.

debt_total_income_category.tail(10)

debt  total_income total_income_category
21515     1        109486                     C
21516     0        322807                     B
21517     0        178059                     C
21518     0        153864                     C
21519     0        115949                     C
21520     0        224791                     B
21521     0        155999                     C
21522     1         89672                     C
21523     1        244093                     B
21524     0         82047                     C

In [44]:
# Провожу анализ зависимости своевременных кредитных обязательств от уровня дохода.

for i in data['total_income_category'].unique():
    print('Доля просрочки по кредиту в категории "',i,'"   составляет:',
    round(100*data['debt'].loc[data['total_income_category'] == i].mean(),1),'%')

Доля просрочки по кредиту в категории " B "   составляет: 7.1 %
Доля просрочки по кредиту в категории " C "   составляет: 8.5 %
Доля просрочки по кредиту в категории " D "   составляет: 6.0 %
Доля просрочки по кредиту в категории " E "   составляет: 9.1 %
Доля просрочки по кредиту в категории " A "   составляет: 8.0 %


Создаю сводную таблицу.

In [45]:
# Создаю таблицу методом pivot().

debt_total_income = data.pivot_table(index='total_income_category', columns='debt', values='gender', aggfunc='count')

In [46]:
# Создаю столбцы таблицы.

debt_total_income.columns = ['no_debt', 'debt']

In [47]:
# Считаю [долю должников].

debt_total_income['share_of_debt'] = debt_total_income['debt'] / (debt_total_income['debt'] + debt_total_income['no_debt'])

In [48]:
# Перевожу значения в столбце [доля дожников] в проценты.

debt_total_income['share_of_debt'] = debt_total_income['share_of_debt'].map('{:.1%}'.format)

In [49]:
# Сортирую по столбцу [доля лолжников] по убыванию.

debt_total_income.sort_values(by='share_of_debt', ascending=False)

no_debt  debt share_of_debt
total_income_category                             
E                           20     2          9.1%
C                        14568  1353          8.5%
A                           23     2          8.0%
B                         4660   354          7.1%
D                          328    21          6.0%

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h5>

Спасибо за интересные наблюдения, на малочисленные категории заемщиков(А, D, E) ориентироваться не стоит, поэтому можно предположить, что более состоятельный клиент - более надежный

</div>

**Вывод:** При рассмотрении данных таблиц, вижу следующий результат: категория граждан "С", наиболее подвержена риску стать должниками по кредитным обязательствам. Однако у данной категории не самый высокий процент по должникам.
Для более точного вывода, нужно использовать репрезентативные выборки категории граждан. Необходимым условием построения репрезентативной выборки является равная вероятность включения в нее каждого элемента множества обьектов, в отношении которых формируется исследовательская гипотеза. Категории E, и A не рассмамтриваю из-за маленькой выборки. Вывод делаю опираясь на результаты категорий C, D и B. Соответственно  при рассмотрении данных категорий вижу, что граждане категории C, имеют наибольший риск стать должниками по кредитным обязательствам и не выплачивать кредиты в срок,это может обьясняться тем что основная масса граждан этой категории имеют не один кредит на руках.Граждане категории D - имеют наименьший риск стать должниками, возможно опасаясь штрафных санкций по просрочкам кредитных обязательств.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [50]:
# Определяю долю несвоевременного возврата кредита в категориях [purpose_category].

for i in data['purpose_category'].unique():
    print('Доля несвоевременного возврата кредита в категории "',i,'"   составляет:',
    round(100*data['debt'].loc[data['purpose_category'] == i].mean(),1),'%')


Доля несвоевременного возврата кредита в категории " операции с недвижимостью "   составляет: 7.3 %
Доля несвоевременного возврата кредита в категории " операции с автомобилем "   составляет: 9.3 %
Доля несвоевременного возврата кредита в категории " получение образования "   составляет: 9.3 %
Доля несвоевременного возврата кредита в категории " проведение свадьбы "   составляет: 7.9 %


Создаю сводную таблицу.

In [51]:
# Создаю таблицу методом pivot_table().

debt_purpose_category = data.pivot_table(index='purpose_category', columns='debt', values='gender', aggfunc='count')

In [52]:
# Создаю столбцы таблицы.

debt_purpose_category.columns = ['no_debt', 'debt']

In [53]:
# Считаю [долю должников].

debt_purpose_category['share_of_debt'] = debt_purpose_category['debt'] / (debt_purpose_category['debt'] + 
                                                                        debt_purpose_category['no_debt'])

In [54]:
# Перевожу значения в столбце [доля дожников] в проценты.

debt_purpose_category['share_of_debt'] = debt_purpose_category['share_of_debt'].map('{:.1%}'.format)

In [55]:
# Сортирую по столбцу [доля лолжников] по убыванию.

debt_purpose_category.sort_values(by='share_of_debt', ascending=False)

no_debt  debt share_of_debt
purpose_category                                     
операции с автомобилем       3879   400          9.3%
получение образования        3619   369          9.3%
проведение свадьбы           2130   183          7.9%
операции с недвижимостью     9971   780          7.3%

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h5>

Хорошее обоснование, ипотечники являются самой обширной категорией и самой добросовестной

</div>

**Вывод:** Граждане берущие кредит для 'операций с недвижимостью', более ответственно подходят к обязательствам по его погашению. категория граждан с кредитами на 'получение образования' и 'операции с автомобилями', больше всего подвержены риску прописаться в категории должников.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причинами могут быть как технические проблемы так и человеческий фактор. Например повреждение дата файла присохранении, или заемщик прост оне захотел раскрывать информацию о своих доходах и трудовой деятельности.
Заполнение информации в датасет могло идти из разных источников, возможно в некоторых случаях информация о доходах была не нужна, например при выдачи кредита участникам зарплатных карт, банка кредитора, и тд.

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h5>

Да, основные причины — технические ошибки и человеческий фактор

</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианное значение лучше всего отражает выборку в целом, в случае, если есть незаполненные данные. Среднее арифметическое отражает ситуацию некорректно, так как не учитывает разброс данных.

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">Комментарий ревьюера (зеленый)</font></h5>

Верно, медиана в отличие от среднего значения, менее чувствительна к выбросам, среднее значение смещается в сторону выбросов.

</div>

### Шаг 4: общий вывод.

 

**Проведенные исследования по предоставленой статистике о платёжеспособности клиентов выявили некоторые зависимости:**    
* **Наличие и количество детей**  

    При анализе данных, выявлен максимальный % не возврата кредита у многодетных клиентов с 4 детьми который составляет 8.9%.  
    При количестве одного и двух детей процент невозврата практически одинаковый (9,2% и 9,4% соответственно).  
        
> Должников с детьми больше, чем бездетных, cледовательно, наличие детей негативно влияет на  
    своевременный возврат кредитных обязательств.
* **Семейное положение**  
    При анализе данных, выявлен максимальный % не возврата кредита у граждан с семейным положением "Не женат / не замужем" и  
    "гражданский брак" (9,8% и 9,3% соответственно).  
    У клиентов с семейным положением "вдовец / вдова" самая низкий % не возврата - 6,6%.  
    
    _Доля просрочки по кредиту в категориях:_  
         1. Не женат / не замужем   = 9.8%
         2. гражданский брак	    = 9.3%
         3. женат / замужем	     = 7.6%
         4. в разводе               = 7.1%
         5. вдовец / вдова          = 6.6%

> Чаще всего должниками оказываются граждане, которые находятся в гражданском браке или в категории не женаты\не замужем.
* **Уровень дохода**  
    При анализе данных, выявлен максимальный % не выполнения кредитных обязательств у клиентов с уровнем дохода  
    категории E (0–30000) - 9.1%.  
    У граждан с уровнем дохода категории A (выше 1 000 001) и C (50 001–200 000)  
    приблизительно одинаковый % не выполнения кредитных обязательств (8% и 8,5% соответственно).   
    Самый низкий % не возврата у клиентов с уровнем дохода категории D (30 001–50 000) - 6.0%.
    
> Зависимость между уровнем дохода и наличием просрочки по кредиту выражена в отношении: кто зарабатывает выше среднего, возвращают долги чаще, чем остальные.
* **Цель кредита**  
    При анализе данных, выявлен максимальный % не выполнения кредитных обязательств у клиентов с целью кредита "операции с  
    автомобилем" и "получение образования" (9,4% и 9,3% соответственно).  
    Самый низкий % не возврата у клиентов с целью кредита "операции с недвижимостью"- 7,3%.   
    
    _Доля несвоевременного возврата кредита в категориях_
    
      1. операции с автомобилем	  = 9.3%  
      2. получение образования	   = 9.3%  
      3. проведение свадьбы	      = 7.9%  
      4. операции с недвижимостью    = 7.3%  
      
> Значительная доля должников брала кредит на автомобиль или образование, граждан, берущих кредит на свадьбу или недвижимость,  меньше.
    
   Граждане входящие в категории многодетных, неженатые и состоящие в гражданском браке с низким уровнем дохода, с наибольшей вероятностью допустят просрочку по кредитным обязательствам.  
   Возможно целью кредита будет получение образования и операции с автомобилем. 
   Комбинация этих факторов может привести к просрочкам. Высокой вероятностью выполнения кредитных обязательств обладают категории граждан, имеющих высокий доход, (выше 200000), и не имеющих детей.
    
   
    

<div class="alert alert-danger"; style="border-left: 7px solid FireBrick">
<h5> ❌ <font color="FireBrick">Комментарий ревьюера (красный)</font></h5>

В общем выводе хочется тезисно видеть все промежуточные выводы с цифрами. Это очень важно, поскольку топ-менеджеры, в большинстве своем, смотрят сразу сюда и только сюда...

Тут важно обратиться к целям, которые были озвучены в самом начале и ответить на все вопросы. Важно оформить все тезисно и с цифрами. Еще важно описать все промежуточные выводы, которых не прописано в целях.
Вкратце структура такая:
- как были предобработаны данные(тут важно написать про изменения, если заполнялись пропуски или преобразовывались строки)
- пройдись по целям и ответь на вопросы
- добавь промежуточные выводы, которые считаешь нужным отметить

Еще очень важно помнить, для кого ты делаешь проект, в зависимости от этого нужно сделать определенные акценты. В данном случае проект учебный, поэтому нужно ориентироваться по описанию.

Для красивого оформления можешь воспользоваться [Кратким руководством по Маркдауну](https://paulradzkov.com/2014/markdown_cheatsheet/).
    

</div>

<div class="alert alert-info"; style="border-left: 7px solid SteelBlue">
<h5>🎓 <font color="SteelBlue">Комментарий от студента (синий)</font></h5>    

Общий вывод сделан!

</div>

<div class="alert alert-success"; style="border-left: 7px solid green">
<h5> ✔️ <font color="green">v2 Комментарий ревьюера (зеленый)</font></h5>

✅ Спасибо, очень хороший и подробный общий вывод, которые содержит все промежуточные результаты 👍
 
</div>